### PR TITLE
fix(behavior_path_planner): pull over deceleration

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

problem
  - Embedded unfeasible rapid deceleration speed with pull over. 
    - this causes overshoots to velocity 0 with smoother

for solve this, improve pull over deceleration
- insert deceleration velocity considering jerk
- request execution module with distance considering current velocity

--------------- 
for more details

- Assumptions
  - Decelerate to the specified speed shortly before the start of the goal search area.
  - Put out blinker after approval
  - To keep the 3-second blinker constraint, embed a speed of `v={distance to parking start}/3s` at each path point.
  - After approval, move to the start point at the current speed.

- before
  - Calculate stopping distance from declaration only
  - When decelerating, linear deceleration speed is embedded for each point, but it does not take into account whether the speed can be decelerated within the constraints.
    - There is a possibility of embedding a speed that requires rapid deceleration near the current position.
  - Module is activated before the fixed distance(100m by default) of the goal
    - ex) When driving at 60 km/h, the module is activated 100 m before the goal, and the goal is modified 20 m before the goal, etc., deceleration is not performed in time.

- after
  - Calculate stopping and decelerating distance considering jerk
  - Embed target speed in deceleration for blinker, compared to minimum deceleration distance for each path point, whichever is farther away.
    - (The possibility of not being able to keep the 3 seconds somewhat comes into play.)
  - For deceleration up to the goal search area, embed the deceleration speed only at the single point where you want to decelerate.
  - The module is activated before the "variable distance with respect to the speed considering the stopping distance".

## Related links

[teir4 internal link
](https://tier4.atlassian.net/browse/T4PB-27401)
## Tests performed

launcher: https://github.com/autowarefoundation/autoware_launch/pull/273

[tier4 internal scenario test](https://evaluation.tier4.jp/evaluation/reports/7a913b59-e9a0-588e-98c2-cfa68f7f7e38?project_id=prd_jt)
1197/1199 -> 1197/1199 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp [hunk 1]
```diff
@@ -151,13 +152,22 @@ class PullOverModule : public SceneModuleInterface
...
   ParallelParkingParameters getGeometricPullOverParameters() const;
+  double calcSignedArchLengthFromEgo(const PathWithLaneId & path, const Pose & pose) const;
+  void decelerateForTurnSignal(const Pose & stop_pose, PathWithLaneId & path) const;
+  void decelerateBeforeSearchStart(const Pose & search_start_pose, PathWithLaneId & path) const;
   std::pair<double, double> calcDistanceToPathChange() const;
   PathWithLaneId generateStopPath();
-  PathWithLaneId generateEmergencyStopPath();
+  PathWithLaneId generateFeasibleStopPath();
+  boost::optional<double> calcFeasibleDecelDistance(const double target_velocity) const;
+  double calcModuleReuquestlength() const;
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp:170]:
```suggestion
  double calcModuleRequestlength() const;
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 2]
```diff
@@ -327,7 +329,7 @@ bool PullOverModule::isExecutionRequested() const
   }
   const double self_to_goal_arc_length =
     util::getSignedDistance(current_pose, goal_pose, current_lanes);
-  if (self_to_goal_arc_length > parameters_->request_length) {
+  if (self_to_goal_arc_length > calcModuleReuquestlength()) {
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:332]:
```suggestion
  if (self_to_goal_arc_length > calcModuleRequestlength()) {
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 3]
```diff
@@ -370,6 +372,19 @@ bool PullOverModule::isExecutionReady() const
   return true;
 }
 
+double PullOverModule::calcModuleReuquestlength() const
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:375]:
```suggestion
double PullOverModule::calcModuleRequestlength() const
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 4]
```diff
@@ -1027,6 +1012,153 @@ TurnSignalInfo PullOverModule::calcTurnSignalInfo() const
...
+    return 0.0;
+  }
+
+  const auto min_stop_distance = calcDecelDistWithJerkAndAccConstraints(
+    v_now, target_velocity, a_now, -a_lim, j_lim, -1.0 * j_lim);
+
+  return min_stop_distance;
+}
+
+double PullOverModule::calcSignedArchLengthFromEgo(
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:1102]:
```suggestion
double PullOverModule::calcSignedArcLengthFromEgo(
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 5]
```diff
@@ -1027,6 +1012,153 @@ TurnSignalInfo PullOverModule::calcTurnSignalInfo() const
...
+{
+  const double time = planner_data_->parameters.turn_signal_search_time;
+  const Pose & current_pose = planner_data_->self_odometry->pose.pose;
+
+  for (auto & point : path.points) {
+    const auto distance_to_stop = std::max(
+      0.0, calcSignedArcLength(path.points, point.point.pose.position, stop_pose.position));
+    const float decel_vel =
+      std::min(point.point.longitudinal_velocity_mps, static_cast<float>(distance_to_stop / time));
+    const double distance_from_ego = calcSignedArchLengthFromEgo(path, stop_pose);
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:1125]:
```suggestion
    const double distance_from_ego = calcSignedArcLengthFromEgo(path, stop_pose);
```

File: planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp [hunk 6]
```diff
@@ -1027,6 +1012,153 @@ TurnSignalInfo PullOverModule::calcTurnSignalInfo() const
...
+void PullOverModule::decelerateBeforeSearchStart(
+  const Pose & search_start_pose, PathWithLaneId & path) const
+{
+  const double pull_over_velocity = parameters_->pull_over_velocity;
+  const Pose & current_pose = planner_data_->self_odometry->pose.pose;
+
+  // slow down before the search area.
+  const auto min_decel_distance = calcFeasibleDecelDistance(pull_over_velocity);
+  if (min_decel_distance) {
+    const double distance_to_search_start = calcSignedArchLengthFromEgo(path, search_start_pose);
```
Comment (rej55) [inline planning/behavior_path_planner/src/scene_module/pull_over/pull_over_module.cpp:1155]:
```suggestion
    const double distance_to_search_start = calcSignedArcLengthFromEgo(path, search_start_pose);
```

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp [hunk 7]
```diff
@@ -151,13 +152,22 @@ class PullOverModule : public SceneModuleInterface
...
 
+  // approximate distance from the start point to the end point of pull_over.
+  // this is used as an assumed value to decelerate, etc., before generating the actual path.
+  double approximate_pull_over_distance_ = 20.0;
+
   bool incrementPathIndex();
   PathWithLaneId getCurrentPath() const;
   Pose calcRefinedGoal(const Pose & goal_pose) const;
   ParallelParkingParameters getGeometricPullOverParameters() const;
+  double calcSignedArchLengthFromEgo(const PathWithLaneId & path, const Pose & pose) const;
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_over/pull_over_module.hpp:163]:
```suggestion
  double calcSignedArcLengthFromEgo(const PathWithLaneId & path, const Pose & pose) const;
```

**Review by rej55 — COMMENTED** [review]
I added some typo-fix suggestions.
Could you check them?

**Comment by kosuke55** [comment]
oh, thanks!

**Review by rej55 — APPROVED** [review]
Sorry for late reaction.
LGTM.

**Pull request merged**